nx/lib/nx/binary_backend.ex

```diff
@@ -476,8 +476,101 @@ defmodule Nx.BinaryBackend do
   end
 
   @impl true
-  def dot(out, %{type: t1} = left, axes1, %{type: t2} = right, axes2) do
-    bin_zip_reduce(out, left, axes1, right, axes2, 0, fn lhs, rhs, acc ->
+  def dot(out, left, contract_axes1, [], right, contract_axes2, []) do
+    # dot/4 is directed to this specific clause so we can keep a more efficient implementation
+    # for non-batched dot products. See the clause below for batched dot products
+    dot4(out, left, contract_axes1, right, contract_axes2)
+  end
+
+  def dot(
+        out,
+        %{shape: left_shape, type: {_, left_size}, names: left_names} = left,
+        left_contract_axes,
+        left_batch_axes,
+        %{shape: right_shape, type: {_, right_size}, names: right_names} = right,
+        right_contract_axes,
+        right_batch_axes
+      ) do
+    left_binary = to_binary(left)
+    right_binary = to_binary(right)
+
+    left_batch_contract_axes = Nx.Shape.shift_axes(left_contract_axes, -length(left_batch_axes))
+
+    right_batch_contract_axes =
+      Nx.Shape.shift_axes(right_contract_axes, -length(right_batch_axes))
+
+    {left_batch_shape, left_batch_names} =
+      Nx.Shape.contract(left_shape, left_batch_axes, left_names, false)
+
+    {right_batch_shape, right_batch_names} =
+      Nx.Shape.contract(right_shape, right_batch_axes, right_names, false)
+
+    left_batch_size = Nx.size(left_batch_shape)
+    right_batch_size = Nx.size(right_batch_shape)
+
+    left_batch_count = Nx.Shape.batch_count(left_shape, left_batch_axes)
+    right_batch_count = Nx.Shape.batch_count(right_shape, right_batch_axes)
+
+    {batch_shape_out, _output_names} =
```

Comment:
I think you should have the output shape in `out`

Reply:
Perhaps the name is wrong. This is the shape for each batch item, not the shape for the batch itself

Reply:
Any suggestions for the name?

Reply:
I changed it to `batch_item_output_shape`